File: Lab-1/Example2/MultithreadingExample.java


public class MultithreadingExample{


    public static void main(String[] args) throws InterruptedException{

        long start = System.nanoTime(); // initial time
        performTask(2000);
        long end = System.nanoTime(); // end time

        System.out.printf("Sequential time: %.3f s%n",(end-start)/1_000_000_000.0);

        start = System.nanoTime();

        Thread thread = new Thread( () -> performTask(2000));
        thread.start();
        thread.join();

        end = System.nanoTime();
        System.out.printf("One-thread time: %.3f s%n",(end-start)/1_000_000_000.0);

        
    }

    public static void performTask(long milliseconds) {
        try {
            Thread.sleep(milliseconds);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }


}

File: Lab-1/Example2/MultithreadingExample.cpp
#include <iostream>
#include <thread>
#include <chrono>

void performTask(long milliseconds){

    std::this_thread::sleep_for(std::chrono::milliseconds(milliseconds));
}

int main(){
    // Sequential execution
    auto start = std::chrono::steady_clock::now();
    performTask(2000);
    auto end = std::chrono::steady_clock::now();
    std::chrono::duration<double> seqElapsed = end - start;
    std::cout << "Sequential time: " << seqElapsed.count() << " s" << std::endl;

    // Execution using one additional thread
    start = std::chrono::steady_clock::now();
    std::thread t(performTask, 2000);
    t.join();
    end = std::chrono::steady_clock::now();
    std::chrono::duration<double> threadElapsed = end - start;
    std::cout << "One-thread time: " << threadElapsed.count() << " s" << std::endl;


    return 0;
}